Make categories visible

shubhamkorde opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8618.
2. This PR does the following: [Makes the category selection dropdown visible on the smaller screen sizes.]

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![oppia_pixel_ss](https://user-images.githubusercontent.com/56932243/106607536-6c7d2d80-6589-11eb-9864-892b2010b5fc.png)
![oppia_ipad_ss](https://user-images.githubusercontent.com/56932243/106607544-71da7800-6589-11eb-9205-234ad0188cb5.png)
![oppia_large_ss](https://user-images.githubusercontent.com/56932243/106607555-756dff00-6589-11eb-951e-e795e0a9b7f4.png)

https://user-images.githubusercontent.com/56932243/106607648-8f0f4680-6589-11eb-8d7a-11be694cadc6.mp4

![Screenshot 2021-02-02 13:20:22](https://user-images.githubusercontent.com/56932243/106607685-9cc4cc00-6589-11eb-88d5-1b81040e35ce.png)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

shubhamkorde committed:
Fix for #8618- Category selection button should be visible on smaller screens

shubhamkorde committed:
Reordered properties in alphabetical order to prevent linting failure

shubhamkorde commented:
https://user-images.githubusercontent.com/56932243/106396886-aa1a7300-6430-11eb-908b-4d26fe20ca9b.mp4

shubhamkorde commented:
@kevintab95 @iamprayush PTAL!

iamprayush commented:
Thanks for the PR @shubhamkorde! I think we should increase the width of the categories dropdown to match the search bar when the dropdown moves below the search bar. Also, @mschanteltc, do you have any thoughts re this?

shubhamkorde commented:
Sure! I will make those changes. Also, I have observed that language selection option vanishes as well when the screen size reduces. Is that the required behaviour?

iamprayush commented:
> Sure! I will make those changes. Also, I have observed that language selection option vanishes as well when the screen size reduces. Is that the required behaviour?

Oh yes, you're right! That's not the required behavior.
Perhaps they could both come under the search bar and occupy 50-50 space, or 60-40, whatever looks better.

shubhamkorde committed:
Made the languages selector option visible

shubhamkorde commented:
@iamprayush PTAL! I have kept some padding between the buttons as it looked better. Is it fine now?

https://user-images.githubusercontent.com/56932243/106499746-f88a4900-64e6-11eb-8b51-e710bc170df6.mp4

iamprayush commented:
I think this looks good, thanks @shubhamkorde!

mschanteltc commented:
This looks great so far! Is it possible that when the dropdown is below the search bar, their borders are in a way combined? Also is it possible to add headings to the dropdown? I added an image for your reference:
![Search Bar](https://user-images.githubusercontent.com/22177674/106520878-46289680-6492-11eb-99a2-93aa808d6a95.png)

shubhamkorde commented:
![Screenshot 2021-02-02 13:21:16](https://user-images.githubusercontent.com/56932243/106568692-9f0f3200-6559-11eb-8a32-d66ca7649579.png)
I tried my best to make it look alike. There is one minor issue, I did not convert "All Categories" text to "All" because on larger screens it makes no sense what that word "All" means.

shubhamkorde committed:
Made the suggested changes as per uploaded image

shubhamkorde committed:
Fixing linting issue

shubhamkorde committed:
Fixed linting issue

shubhamkorde commented:
@iamprayush @mschanteltc PTAL!

iamprayush commented:
Hey @shubhamkorde, could you please add a video/screenshots that show all the different screen sizes, in the "**Proof that the changes are correct**" section of the PR description?

shubhamkorde commented:
@iamprayush I have added the video as well as screeshots in the proofs section. PTAL!

mschanteltc commented:
> I tried my best to make it look alike. There is one minor issue, I did not convert "All Categories" text to "All" because on larger screens it makes no sense what that word "All" means.

Thanks this looks great! Is it possible to apply headings to the dropdown menu in all formats to avoid that confusion? Or is the dropdown menu's height too tall for the navbar?

shubhamkorde commented:
It is too tall for for large screens desktops and landscape tablets. If I try including heading/label in all formats without increasing the navbar height it will become to small to read (lesser than 8px font size). If we go ahead and try to change the navbar height for this page it will not look uniform across all pages.
The possible things that I can think we can do are:
**1) Remove the language and category selection option from the nav-bar for all screen sizes.**
Pros: No space crunch. We can put in the labels as we want. We can also improve the readability, styling etc.
Cons: The nav-bar might start looking really empty. We will have to redo a lot of styling that already exists (which looks great now!)

**2) Add hover effect**
When the user hovers over the dropdown the text should change to what it represents. For example: The text "English" could be changed to "Select Language" when someone hovers over it. 
Pros: Easy fix. We will not have to change the styling for smaller screens.
Cons: In a single glance user cannot make out what it represents. Hovering will be necessary. Also hover effect might seem out of place as it has not been used in the rest of the website.

**3) Just leave it as it is** as "All Categories" and "English" are quite representative of the button might do. Most users won't have a big problem figuring out. 

What should I do? @mschanteltc

mschanteltc commented:
@shubhamkorde I'm thinking we should leave the search bar in the navbar (if the format allows it) for the same reason you mentioned which is the navbar would look empty without it. Would it be possible though we can add an "Advanced Search" feature to the side of the search bar? Clicking it would navigate to a page similar to [this](https://cdn.dribbble.com/users/1798715/screenshots/6066303/advanced_search_4x.jpg?compress=1&resize=400x300) where users can specify and narrow down their search.

shubhamkorde commented:
Yes, that will be great. I would like to work on that part as well. Will that be a part of this issue or will you be opening a new issue? Also, I don't know why the tests are failing :(. Do you have any idea why is this happening? @mschanteltc @iamprayush. 

Fixed it. Thanks!

shubhamkorde committed:
removed additional spacing

shubhamkorde commented:
@kevintab95 @mschanteltc @iamprayush tests have passed now. PTAL!

mschanteltc commented:
> @kevintab95 @mschanteltc @iamprayush tests have passed now. PTAL!

Thanks @shubhamkorde. Where can I view the changes?

shubhamkorde commented:
@mschanteltc I have added the videos in the description of the PR of how it looks. The code changes could be seen in the subheading labelled "Files changed". You can also replicate these changes locally on your machine via git or simply copy pasting the single file that I have changed.

iamprayush commented:
LGTM, thanks @shubhamkorde!

DubeySandeep commented:
@kevintab95, Can you please review this PR?

kevintab95 reviewed (approved):
LGTM, thanks @shubhamkorde!

mschanteltc commented:
LGTM
